fn/goroutine_manager.go

```diff
@@ -56,20 +65,25 @@ func (g *GoroutineManager) Go(f func(ctx context.Context)) bool {
 // Stop prevents new goroutines from being added and waits for all running
 // goroutines to finish.
 func (g *GoroutineManager) Stop() {
-	g.mu.Lock()
-	g.cancel()
-	g.mu.Unlock()
+	g.stopped.Do(func() {
+		// Closing the quit channel will cause all goroutines to exit
+		// and will prevent any new goroutines from starting.
+		g.mu.Lock()
+		close(g.quit)
+		g.mu.Unlock()
 
-	// Wait for all goroutines to finish. Note that this wg.Wait() call is
-	// safe, since it can't run in parallel with wg.Add(1) call in Go, since
-	// we just cancelled the context and even if Go call starts running here
-	// after acquiring the mutex, it would see that the context has expired
-	// and return false instead of calling wg.Add(1).
-	g.wg.Wait()
+		// Wait for all goroutines to finish. Note that this wg.Wait()
+		// call is safe, since it can't run in parallel with wg.Add(1)
+		// call in Go, since we just cancelled the context and even if
+		// Go call starts running here after acquiring the mutex, it
+		// would see that the context has expired and return false
+		// instead of calling wg.Add(1).
+		g.wg.Wait()
```

Comment:
is that a problem? is there a usecase for needing to support the blocking behaviour? afaiu, this will mostly be used within other sub-systems and Stop will be called by their Stop methods which typically will also only be called once

Reply:
I just dont think more than 1 system will ever own the GoroutineManager right? 

ie, what makes this Stop different from other Stop methods of other subsystems in LND which use a similar pattern to this?

Reply:
but why will there ever be a second Stop call is what im getting at. Look at all the instances of `) Stop() {` in the codebase

Reply:
I think that if we really want to have a method that explicitly blocks and does so for multiple routines, then we should have an explicit `Done() <-chan struct` method that does this

Reply:
i see there is a Done() method - we could always update this to return a dedicated `shutdown` channel that is only closed after Wait() in Stop

Reply:
@bhandras - perhaps a 3rd opinion here just to break the tie would help? 🙏

Reply:
Option 3 sgtm!

Reply:
cool - pushed a diff with option 3 included 🫡

Reply:
ah - just realised we cannot do this since callers may wait on Done() inside a goroutine that was started within Go().

Undoing this change